python/taichi/ad/_ad.py

```diff
@@ -7,16 +7,134 @@
 from taichi.lang.enums import SNodeGradType
+from taichi.lang.expr import Expr
 from taichi.lang.field import ScalarField
+from taichi.lang.kernel_impl import kernel
 from taichi.lang.snode import SNode
+from taichi.types import ndarray, template
 
 from taichi import _snode
 
 
+class GradChecker:
+    def __init__(self, loss, to_check):
+        self.to_check = to_check
+        self.loss = loss
+        self.eps_range = 2.**np.arange(-3, -30, -2).astype(np.float64)
+        self.result = [None] * len(to_check)
+        self.all_fields = get_all_fields()
+        self.backups = save_all_fields(self.all_fields)
+
+    def add_calls(self, calls):
+        self.calls = calls
+
+    def check_grad(self):
+        assert self.loss.dtype == types.f64, "Only f64 is supported when checking grad."
+
+        @kernel
+        def x_pos(x: template(), tangent_np: ndarray(), eps: types.f64):
+            for I in impl.grouped(x):
+                x[I] += eps * tangent_np[I]
+
+        @kernel
+        def x_neg(x: template(), tangent_np: ndarray(), eps: types.f64):
+            for I in impl.grouped(x):
+                x[I] -= eps * tangent_np[I]
+
+        for i, x in enumerate(self.to_check):
+            if x is self.loss:
+                self.result[i] = True
+                continue
+
+            check_pass = False
+
+            re_range = []
+            for eps in self.eps_range:
+                tangent_np = np.array(np.random.rand(*x.shape)).astype(
+                    np.float64)
+
+                restore_all_fields(self.all_fields, self.backups)
+                x_pos(x, tangent_np, eps)
+                for func, args in self.calls:
+                    func(*args)
+                loss_pos = self.loss.to_numpy()
+
+                restore_all_fields(self.all_fields, self.backups)
+                x_neg(x, tangent_np, eps)
+                for func, args in self.calls:
+                    func(*args)
+                loss_neg = self.loss.to_numpy()
+
+                ip_numerical = (loss_pos - loss_neg) * 0.5 / eps
+                x_grad_np = x.grad.to_numpy()
+                extra_dim = x_grad_np.ndim - tangent_np.ndim
+                if extra_dim == 1:
+                    tangent_np = np.expand_dims(tangent_np, axis=-1)
+                if extra_dim == 2:
+                    tangent_np = np.expand_dims(tangent_np, axis=-1)
+
+                ip_autodiff = np.sum(x_grad_np * tangent_np)
+                err = abs(ip_autodiff - ip_numerical)
+                if ip_numerical != 0:
+                    re = err / abs(ip_autodiff)
+                else:
+                    re = err / (abs(ip_autodiff) + 1e-20)
+                re_range.append(re)
+
+                if err * 100 <= abs(ip_autodiff):
+                    check_pass = True
+                    break
+
+            self.result[i] = check_pass
+
+            if not check_pass:
+                print("variable", i, "has relative error", re_range)
```

Comment:
nit: can we make it a warning e.g., `Grad check failed: variable xxx has relative error xxx, expected relative error xxx `

Reply:
okay！